entity-data-service-rx-client/src/main/java/org/hypertrace/entity/data/service/rxclient/EntityDataCachingClient.java

```diff
@@ -0,0 +1,44 @@
+package org.hypertrace.entity.data.service.rxclient;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import io.grpc.CallCredentials;
+import io.grpc.Channel;
+import io.reactivex.rxjava3.core.Single;
+import java.time.Duration;
+import javax.annotation.Nonnull;
+import org.hypertrace.entity.data.service.v1.Entity;
+import org.hypertrace.entity.data.service.v1.EntityDataServiceGrpc;
+import org.hypertrace.entity.data.service.v1.EntityDataServiceGrpc.EntityDataServiceStub;
+
+class EntityDataCachingClient implements EntityDataClient {
+  private final EntityDataServiceStub entityDataClient;
+  private final LoadingCache<EntityCacheKey, Single<Entity>> cache;
+
+  EntityDataCachingClient(
+      @Nonnull Channel channel,
+      @Nonnull CallCredentials credentials,
+      int maxCacheContexts,
+      @Nonnull Duration cacheExpiration) {
+    this.entityDataClient = EntityDataServiceGrpc.newStub(channel).withCallCredentials(credentials);
+    this.cache =
+        CacheBuilder.newBuilder()
+            .maximumSize(maxCacheContexts)
+            .expireAfterWrite(cacheExpiration)
+            .build(CacheLoader.from(this::upsertEntity));
+  }
+
+  @Override
+  public Single<Entity> getOrCreateEntity(Entity entity) {
```

Comment:
mmh maybe we need to clarify that the caching is based on (type, id, tenantId). There are other case where we get by identifying attributes and that would need a different kind of cache key.

Reply:
I can clarify in the docs - the concept of identifying attributes is going away in the new entity model.

Reply:
Will need to hold this PR for a bit - the upsert API makes some assumptions that don't work with the new entity model.

Reply:
Handled in #41

Reply:
I just realized I forgot to actually update the docs there. update incoming.

Reply:
Ok you win, when writing the docs I felt like this is prone to be used with a v1 entity and cause strange bugs. Will add support for either type in the key.